sync-global-pullsecret/sync-global-pullsecret.go

```diff
@@ -0,0 +1,257 @@
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	hyperapi "github.com/openshift/hypershift/support/api"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/coreos/go-systemd/dbus"
+	"github.com/spf13/cobra"
+)
+
+// syncGlobalPullSecretOptions contains the configuration options for the sync-global-pullsecret command
+type syncGlobalPullSecretOptions struct {
+	kubeletConfigJsonPath string
+	globalPSSecretName    string
+}
+
+const (
+	defaultKubeletConfigJsonPath     = "/var/lib/kubelet/config.json"
+	defaultGlobalPSSecretName        = "global-pull-secret"
+	defaultGlobalPullSecretNamespace = "kube-system"
+	dbusRestartUnitMode              = "replace"
+	kubeletServiceUnit               = "kubelet.service"
+
+	// systemd job completion state as documented in go-systemd/dbus
+	systemdJobDone = "done" // Job completed successfully
+)
+
+//go:generate ../hack/tools/bin/mockgen -destination=sync-global-pullsecret_mock.go -package=syncglobalpullsecret . dbusConn
+type dbusConn interface {
+	RestartUnit(name string, mode string, ch chan<- string) (int, error)
+	Close()
+}
+
+// writeFileFunc is a variable that holds the function used to write files.
+// This allows tests to inject custom write functions for testing rollback scenarios.
+var writeFileFunc = os.WriteFile
+
+// GlobalPullSecretReconciler reconciles a Secret object
+type GlobalPullSecretReconciler struct {
+	client.Client
+	Scheme                *runtime.Scheme
+	kubeletConfigJsonPath string
+	globalPSSecretName    string
+	globalPSSecretNS      string
+}
+
+// NewRunCommand creates a new cobra.Command for the sync-global-pullsecret command
+func NewRunCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "sync-global-pullsecret",
+		Short: "Syncs a mixture between the user provided pull secret in DataPlane and the HostedCluster PullSecret to be deployed in the nodes of the HostedCluster",
+		Long:  `Syncs a mixture between the user provided pull secret in DataPlane and the HostedCluster PullSecret to be deployed in the nodes of the HostedCluster. The resulting pull secret is deployed in a DaemonSet in the DataPlane that updates the kubelet.config.json file with the new pull secret. If there are conflicting entries in the resulting global pull secret, the user provided pull secret will prevail.`,
+	}
+
+	opts := syncGlobalPullSecretOptions{
+		kubeletConfigJsonPath: defaultKubeletConfigJsonPath,
+	}
+	cmd.Flags().StringVar(&opts.globalPSSecretName, "global-pull-secret-name", defaultGlobalPSSecretName, "The name of the global pullSecret secret in the DataPlane.")
```

Comment:
why does this global-pull-secret-name need to be a param?

Reply:
for flexibility, if we need to change the secret name, it's matter of pass that name to the controller

Reply:
the name should never change and if it's does, how is a param more flexible than a hardcoded value providing the ds instantiation is codified itself? The only value I can think of is to run the daemon set manually for dev purposes